### PR TITLE
convai2 task

### PR DESCRIPTION
**Pull request by emilydinan**
Task for ConvAI2 challenge.

Data is taken from personachat dataset with some edits:
1. Contractions are added back in ~50% of the time (i.e. "I am" is sometimes changed to "I'm")
2. Data is doubled: to each file, we append the dataset again where the first utterance in the conversation is `<SILENCE>` and then the previous first utterance becomes the label, etc. Personas are switched in the appended data so that they make sense. We use the candidates from the original data, but switch out the true label in the list of candidates to the correct one.

### CODE REVIEW

**Comment by jaseweston** [comment]
we should probably make totally random candidates, even if they are not used usually in training.
for valid and test we don't need to double.

**Comment by emilydinan** [comment]
@jaseweston Latest version on AWS is such that 
- candidates for "appended data" are pulled randomly from the training set, excluding the conversation. we check to make sure the label is not duplicated within the candidate set.
- only train sets are doubled (valid and test are left as is from personachat, but with contractions sometimes replaced as described above)

File: parlai/tasks/convai2/agents.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+class SelfTeacher(SelfOriginalTeacher):
+    pass
+
+class SelfRevisedTeacher(FbDialogTeacher):
+    def __init__(self, opt, shared=None):
+        opt = copy.deepcopy(opt)
+        opt['datafile'] = _path(opt, 'self_revised')
+        super().__init__(opt, shared)
+
+class OtherOriginalTeacher(FbDialogTeacher):
```
Comment (jaseweston) [inline parlai/tasks/convai2/agents.py]: oh i think we should get rid of the other and both actually for the competition

**Comment by alexholdenmiller** [comment]
can we do __SILENCE__ instead to match e.g. our dictionary class's special tokens?

File: parlai/tasks/convai2/agents.py (hunk 1 above)
Comment (alexholdenmiller) [reply parlai/tasks/convai2/agents.py]: +1 on this

**Comment by emilydinan** [comment]
- Changed `<SILENCE>` token to `__SILENCE__`
- Got rid of "other" and "both" teachers

**Comment by emilydinan** [comment]
@jaseweston @alexholdenmiller 
- latest version on AWS combines test and validation sets to just be the validation set. `convai/agents.py` edited to issue a warning when you try to use a test set and sets the datatype to "valid."
- Also includes small spelling fix ("cant" was not included in list of original contractions that were fixed; changed to sometimes be "can't").

**Comment by jaseweston** [comment]
super! any thing left to do? if not lets get it in.

**Comment by emilydinan** [comment]
@jaseweston the only minor thing to decide is what we want the default teacher to be -- right now set to selfOriginal but we had discussed setting to selfRevised. this can obviously be changed later if we want to merge now.

**Comment by jaseweston** [comment]
ok, yes let's check in and think about it
